compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/cyclefind/GlobalVariableRefAnalyzer.java

```diff
@@ -0,0 +1,355 @@
+        List<BSymbol> globalVars = this.pkgNode.globalVars.stream()
+                .filter(v -> v.symbol != null)
+                .map(v -> v.symbol)
+                .collect(Collectors.toCollection(ArrayList::new));
+
+        List<BSymbol> funcs = this.globalNodeDependsOn.keySet().stream()
+                .filter(symbol -> !globalVars.contains(symbol))
+                .collect(Collectors.toList());
+
+        funcs.addAll(globalVars);
+
+        return funcs;
+    }
+
+    private void moveAndAppendToSortedList(BSymbol symbol, List<BSymbol> moveFrom, Set<BSymbol> sorted) {
+        sorted.add(symbol);
+        moveFrom.remove(symbol);
+    }
+
+    private int analyzeProvidersRecursively(NodeInfo node) {
+        if (node.visited) {
+            return node.lowLink;
+        }
+
+        node.visited = true;
+        node.lowLink = node.id;
+        node.onStack = true;
+        nodeInfoStack.push(node);
+
+        List<BSymbol> providers = globalNodeDependsOn.getOrDefault(node.symbol, new ArrayList<>());
+        for (BSymbol providerSym : providers) {
+            NodeInfo providerNode =
+                    dependencyNodes.computeIfAbsent(providerSym, s -> new NodeInfo(curNodeId++, providerSym));
+            int lastLowLink = analyzeProvidersRecursively(providerNode);
+            if (providerNode.onStack) {
+                node.lowLink = Math.min(node.lowLink, lastLowLink);
+            }
+        }
+        // Cycle detected.
+        if (node.id == node.lowLink) {
+            List<NodeInfo> cycle = new ArrayList<>();
+
+            while (!nodeInfoStack.isEmpty()) {
+                NodeInfo cNode = nodeInfoStack.pop();
+                cNode.onStack = false;
+                cNode.lowLink = node.id;
+                cycle.add(cNode);
+                if (cNode.id == node.id) {
+                    break;
+                }
+            }
+            cycles.add(cycle);
+            if (cycle.size() > 1) {
+                cycle = new ArrayList<>(cycle);
+                Collections.reverse(cycle);
+                List<BSymbol> symbolsOfCycle = cycle.stream()
+                        .map(n -> n.symbol)
+                        .collect(Collectors.toList());
+
+                if (doesContainAGlobalVar(symbolsOfCycle)) {
+                    emitError(symbolsOfCycle);
+                }
+            }
+        }
+        dependencyOrder.add(node);
+        return node.lowLink;
+    }
+
+    private void emitError(List<BSymbol> symbolsOfCycle) {
+        Optional<BSymbol> firstGlobalVar = symbolsOfCycle.stream().filter(s -> s.kind != SymbolKind.FUNCTION).findAny();
+        if (!firstGlobalVar.isPresent()) {
+            return;
+        }
+
+        BSymbol firstNodeSymbol = firstGlobalVar.get();
+        Optional<BLangNode> firstNode = pkgNode.topLevelNodes.stream()
+                .filter(t -> t.getKind() == NodeKind.VARIABLE || t.getKind() == NodeKind.FUNCTION)
```

Comment:
I think we did this to remove code warning? If not we can use one filter like
t -> (t.getKind() == NodeKind.VARIABLE || t.getKind() == NodeKind.FUNCTION) && getVarOrFuncSymbol((BLangNode) t) == firstNodeSymbol